### ISSUE TITLE
[DASH] Bitstream switching between Representations that use different codecs.

### ISSUE THREAD

**Issue by vladimir-kazakov** [gpac/gpac#23]
Hi!

While creating DASH media using MP4Box (GPAC rev 5572), I noticed a warning that says "[DASH]: Couldnt merge AVC|H264 SPS from different files (same SPS ID used) - different sample descriptions will be used". Also, by looking at the code and by checking the MPD, I noticed that bitstream switching is turned off in this case: https://github.com/gpac/gpac/blob/25a10ffec1af2431a708d3b0ca4133c228cc4787/src/media_tools/dash_segmenter.c#L2437

My media consists of 2 video (v) and 1 audio (a) stream (http://1drv.ms/1xDsl27):
1.v) 360p using Constrained Baseline profile and level 3.0.
2.v) 720p using High profile and level 3.1.
1.a) AAC-LC.

So, what triggers this behavior (and warning) is the fact that I'm using different codecs for different video Representations.

But it seems that something is wrong here.
1) DASH-AVC/264 Interoperability Points v2.0 (and other versions) does not require Representations in one specific AdaptationSet to use the same codec. (This case is not even mentioned there.)
2) ISO/IEC 23009-1 (2nd edition) (DASH standard) also does not have such a requirement, but it additionally mentions in 4.3 that "Clients may ignore Representations that rely on codecs or other rendering technologies they do not support or that are otherwise unsuitable".

In other words, only media players (Clients) must decide whether they're going to switch to specific (supported) streams. So, bitstream switching must not be disabled by default, if different codecs are used.

Also, speaking of the warning itself, it seems wrong or misleading (although I must say that I don't know much about SPS NAL units). Why MP4Box is trying to merge SPSs? Since different profiles and levels are used, it should be OK that SPSs are different, isn't it? And if SPSs "couldn't be merged", then why the "same SPS ID used"?

**Comment by jeanlf** [comment]
The warning was not thrown correctly, this should now be fixed with rev 5583.
The merging of SPS only happens when using -bs-switching merge option, it's a special mode where SPSs from the different inputs are combined, when possible, in a single track description to allow for bitstream switching in 'avc1' mode rather than 'avc3' mode

**Comment by vladimir-kazakov** [comment]
@jeanlf, thanks for fixing it!

But I have some additional questions. -bs-switching by default is set to "inband". So, when I use multiple video streams that use the same codec, in MPD I see "avc1.\*" values in Representation@codecs. But when I use multiple video streams that use different codecs, in MPD I see "avc3.\*" values Representation@codecs. dash.js (1.2.0, and the latest from the development branch) in this case doesn't show the video (black screen), although the playback works and I hear the audio. I've tried to set -bs-switching to "merge". MP4Box reported "[DASH]: Could not merge AVC|H264 SPS from different files (same SPS ID used) - disabling bitstream switching", and in MPD, values in Representation@codecs are again set to "avc1.\*", although the bitstream switching is turned off (this attribute is missing and by default it's set to "false", according to the DASH specification). But at the same time, this version works fine in dash.js. Well, actually, DASH-IF IOP v2.0 says in 4.2.2 (line 20): "Clients are expected to support Inband Storage for SPS/PPS based ISO/IEC 14496-15, 20 Amendment 2 [22], i.e. sample entry 'avc3' and 'avc4'". So, I assume that this is a bug in dash.js. But before reporting it, I'd like to better understand this topic.

So, my questions are:

1. These avc modes (avc1, avc3, etc.). I don't know anything about them, and a quick search on this topic didn't give me any results. Could you recommend me some source of information where these modes are described? Or, perhaps, maybe you could explain it yourself? I'm not looking for low-level details. I just want to know enough in order to produce correct media.
1. Since SPSs can't be merged, does it mean that it's better to avoid media that uses different codecs for different video Representations in the case when bitstream switching is needed? If so, then what are the problems (assuming that they exist) (including potential ones) for a media player to switch between such Representations?

**Comment by jeanlf** [comment]
1) (and same applies to HEVC): AVC and HEVC store codec config in SPS and PPS (and VPS for HEVC). The "codec parameters" can be stored in the sample entry of the track (eg in avc1 or hvc1), in which case they shall not occur in the sample data. Or they can be stored in the sample data (avc3 and hev1) in which case they may not be present in the sample entry; this mode allows for dynamic change of the codec config but complexifies a bit the seeking (need to find the closest previous IDR with the right parameter sets). 

2) no, because most (if not all) players are capable of handling non "bitstream-switching" sequences. The terminology here is a bit confusing: the fact that you don't have bitstreamSwitching="true" for an adaptationSet does not mean the client cannot switch between the representations of that set, it means that the client will have to re-parse a moov (init segment) before the switch.

**Comment by vladimir-kazakov** [comment]
@jeanlf, thanks for clarification! My mistake was that I was thinking about the @bitstreamSwitching in MPD as about an indication that a media player must (or must not) attempt to switch between Representations.

**Comment by vladimir-kazakov** [comment]
@jeanlf, I have more questions related to this topic.

Right now I'm using GPAC 0.5.2-DEV-rev80-g29b60b6-master. I have a video that consists of these Representations:

* 360p; Profile: Constrained Baseline; Level: 3.0; Codec: 42c01e
* 720p; Profile: High; Level 3.1; Codec: 64001f
* 1080p; Profile: High: Level 4.0; Codec: 640028

Questions:

1. When I'm creating a DASH media by using the "dashavc264:**live**" profile, I see that the `bitstreamSwitching="true"`, which is expected, since the default mode is "inband". I also see that the "avc3" mode is used. But what's strange is that all 3 Representations have `codecs="avc3.42c01e"`. I was expecting that the codec part for each Representation must be unique. Is everything correct here?
2. When I'm creating a DASH media by using the "dashavc264:**onDemand**" profile, I see that the `bitstreamSwitching` attribute is missing, and the "avc1" mode is used. I didn't expect this, so I specified `-bs-switching inband`, and I got the same result - nothing has changed. Looks like something is wrong here, but there are not warnings or errors.

**Comment by jeanlf** [comment]
1) this is strange, you should have different codecs params for each representation, could you share your test files somewhere ?

2) check the spec, this is normal: onDemand forces that each representation consists of a single segment self-intialiizing (eg with moov included). Bitstream switching is therefore ignored

**Comment by vladimir-kazakov** [comment]
@jeanlf 

1. http://1drv.ms/1zaXrZZ
2. I guess it would be nice if there will be an info / verbose / warning message that says that some parameter will be ignored for a good reason. Otherwise, it's not clear what's happening.

**Comment by jeanlf** [comment]
thanks for the files. this should now be fixed (also added an error message in case one wants bs switching in ondemand profile)

**Comment by vladimir-kazakov** [comment]
@jeanlf, looks like everything is ok now, but perhaps one thing can be improved.

Right now, the default mode is the "inband". That's ok if codecs are different, but perhaps it's not really a good choice if codecs are the same, since looks like web-browsers don't really support "avc3.*" media very well. I guess, it would be better if MP4Box would look at codecs in all Representations, and, based on this information, it would choose the mode to use (if the mode wasn't explicitly specified). For example, if all Representations use the same codec - the mode is "no". If codecs are different - the mode is "inband".

I'm just thinking about how people will be creating DASH media. Previously they didn't have to think about bitstream switching at all. Now, suddenly, they will start getting "avc3.*" instead of "avc1.*", which will result in https://github.com/Dash-Industry-Forum/dash.js/issues/445.

**Comment by rbouqueau** [comment]
@jeanlf I'm going to conduct tests regarding the multiple sample entries. If it works well in browsers, we could make it the default parameter, what do you think?

**Comment by Murmur** [comment]
Dash multibitrate with -bs-switching multi creates a single init.mp4 file where TRAK/MDIA/MINF/STBL table has three AVC1 children. Child element has SPS and PPS h264 stream details. 

mp4box uses  `MOOF/TRAF/TFHD/sample_description_index` 2 and 3 numbers in an appropriate track fragment files. Fragments for first video track has `MOOF/TRAF/flags=20000` so index field is not explicitly set, it maps to first avc1 child. That's my limited mp4 understanding. mp4box writes manifest `codecs="avc3.4d4028"` video representation attributes.

This test stream from internet has an explicit `MOOF/TRAF/TFHD/sample_description_index` 1, 2 and 3 numbers in all track fragment files. This file is multi_avc1 file but codecs attributes are using avc1.* prefix.
[Where_The_Trail_Ends_768_448_384_ddp.mpd](http://streaming.dolby.com/ondel/playback/Dolby_Digital_Plus_Online_Playback_Kit_v1.1/Test_Signals/HbbTV/HPL40_25fps/Unencrypted/multi_av_rate/71/Where_The_Trail_Ends_768_448_384_ddp.mpd)
[video init file](http://streaming.dolby.com/ondel/playback/Dolby_Digital_Plus_Online_Playback_Kit_v1.1/Test_Signals/HbbTV/HPL40_25fps/Unencrypted/multi_av_rate/71/Where_The_Trail_Ends_768_448_384_ddp_V.mp4), [video1 seg1](http://streaming.dolby.com/ondel/playback/Dolby_Digital_Plus_Online_Playback_Kit_v1.1/Test_Signals/HbbTV/HPL40_25fps/Unencrypted/multi_av_rate/71/Where_The_Trail_Ends_768_448_384_ddp_V_1_1.m4s), [video2 seg1](http://streaming.dolby.com/ondel/playback/Dolby_Digital_Plus_Online_Playback_Kit_v1.1/Test_Signals/HbbTV/HPL40_25fps/Unencrypted/multi_av_rate/71/Where_The_Trail_Ends_768_448_384_ddp_V_2_1.m4s), [video3 seg1](http://streaming.dolby.com/ondel/playback/Dolby_Digital_Plus_Online_Playback_Kit_v1.1/Test_Signals/HbbTV/HPL40_25fps/Unencrypted/multi_av_rate/71/Where_The_Trail_Ends_768_448_384_ddp_V_3_1.m4s)

```
## MP4Box - GPAC version 0.5.2-DEV-rev384-gaa0f405-master
mp4box.exe -noprog -dash 8000 -frag 8000 -mem-frags -rap -frag-rap -profile dashavc264:live
 -profile-ext "urn:hbbtv:dash:profile:isoff-live:2012"
 -bs-switching multi -sample-groups-traf -single-traf -no-frags-default -subsegs-per-sidx 0
 -min-buffer 4000
 -segment-name "testvideo_$RepresentationID$_$Number$$Init=init$" -segment-timeline
 -out "output\testvideo.mpd"
 "temp-360p.mp4#trackID=1:id=v1"
 "temp-720p.mp4#trackID=1:id=v2"
 "temp-1080p.mp4#trackID=1:id=v3"
 "temp-128k.mp4#trackID=1:id=a4"
```

I have access to several HbbTV televisions and all devices playback Dolphy Where_The_Trail_Ends_768_448_384_ddp.mpd test file. It is h264+ec3 multibitrate file.

My own test file works fine on some devices with avc3.* codec prefix. Some failed but after editing avc1.* prefix in codecs attributes manifest worked fine. Some devices don't playback anything. I need to study this further but am gathering random differences so far.
- first video stream fragments don't use an explicit sample_description_index value
- Dolphy has ec3 audio but I had aac_lc, need to change this to aac_he, rip ec3 for testing or don't use audio track at all.
- I had 640x360, 1280x720, 1920x1080 video size, Dolphy has 720x576, 1280x720, 1920x1080 size. Maybe hbbtv devices don't work well with non-broadcast resolutions. I did try single bitrate 360p and it did work on a problematic device.

I don't have aac_he in my ffmpeg build yet so that may have a problem. I need to download non-gpl ffmpeg with libfdk_aac modules and try this difference later.

This problem sort of belongs to this topic but I am not sure if should create a new issue topic, or just communicate this here for a while. Missing explicit sample_index is major(?) difference I have found in meta data.
